### PR TITLE
[scripts][nnet3] Remove the ability to separately specify validation …

### PR DESCRIPTION
**Pull request by danpovey**
…l/r context; use regular chunk-size for valid/train-subset egs.

I'm not sure whether to aim to test this PR separately and merge it soon, or to wait till my other extensive changes to nnet3 stuff, including the egs creation process and the compilation process, need to be tested.

### CODE REVIEW

**Comment by danpovey** [comment]
@vimalmanohar and @vijayaditya, you may want to glance at this, but no need to thoroughly check it, it's straightforward..;

@hhadian, would you mind testing this for me?  No hurry-- the process can take a couple of weeks.  It will be better to get this tested independently, so that later, when we check the more extensive changes to the nnet3 setup, we know that this one is already checked in and is not causing any problems.

The important things to run are the nnet3 LSTM[or, in general, recurrent-RNN] scripts and the discriminative training scripts on top of that.  Take note of how long it takes to compute the train/valid probs on each iteration, and the various diagnostic probs; also the time taken in the 'combine' step.
Best to do this on top of a modern, xconfig-based setup, maybe a TDNN+LSTM setup followed by discriminative training, you can create a discriminative training script for that setup and make a link to the tuning directory for that.
Actually more scripts than these are affected by the changes, but these are the most likely to be affected.
Also check the time taken in the compute-priors jobs in the discriminative training.

**Comment by vijayaditya** [comment]
OK will do.

--Vijay

**Comment by hhadian** [comment]
OK, I will do it.

**Comment by hhadian** [comment]
So I am going to do the following:
1. Prepare a TDNN+LSTM setup for nnet3 using xconfig (I was going to do this anyway; for low frame rate experiments)
2. Prepare a discriminative setup on top of that
3. Try these scripts before and after applying the changes in this PR, and compare the results and runtime of different computations.
Are there any scripts/tools to report/compare the times taken for computing probs, combining, etc. ?
Hossein

**Comment by danpovey** [comment]
Sounds great-- thanks.

There is no special tool but the relevant log files are the
compute_prob_*.log and the 'combine.log'... there should be a line with
'Accounting:' in it that gives you the time taken.

**Comment by hhadian** [comment]
OK. I will get the times from the log files.
Regarding TDNN+LSTM for nnet3, I am preparing it based on chain scripts (local/chain/run_tdnn_lstm.sh) and current nnet3/LSTM script (local/nnet3/run_lstm.sh). These options are different:

`deriv-truncate-margin` --> no such thing in nnet3/train_rnn.py
`momentum` is 0.0 for chain, but it is 0.5 for nnet3/LSTM setup
initial and final learning rate for chain are `0.001, 0.0001` whereas they are `0.0003, 0.00003` for nnet3/LSTM (significant difference)

What should I do about these?

**Comment by danpovey** [comment]
Hm.  The option to train_rnn.py is called
--trainer.deriv-truncate-margin
.. which is the same as in the chain script... make sure your repo is up to
date if not.
As for momentum-- you could try 0.5.  [I think we found that momentum is
not helpful for chain; anyway it was never a huge difference.]
As for the initial/final learning rates: maybe take them from the
nnet3+LSTM setup.  I'm not 100% sure if we have a rule of thumb about how
the learning rates should differ in nnet3 versus chain, @vijayaditya may
know; but anyway the objf has a lower dynamic range for chain so it's not
surprising if we'd want lower learning rates for chain.

**Comment by hhadian** [comment]
Oh! I will make the code up-to-date now.
I will try 0.5 for momentum. 
Since the learning rates for nnet3/TDNN are much larger (0.0017, 0.00017) than nnet3/LSTM (0.0003, 0.00003), and since this setup is a combination of TDNN and LSTM, a value in between might make sense for the start (i.e. 0.001, 0.0001). Is that OK to start with?

**Comment by danpovey** [comment]
No, I'd use the LSTM one.  The TDNN layers are much less sensitive to the
learning rate schedule because they are invariant to the scale of the
parameters [thanks to the renorm layer], so if the learning rate is too
large the parameters just get larger to compensate.

**Comment by hhadian** [comment]
OK I'll go with LSTM's.

**Comment by danpovey** [comment]
I'm closing this PR; the code is in the shortcut branch and will be merged when that branch is merged.
